tests/attacks/test_pixel_attack.py

```diff
@@ -0,0 +1,174 @@
+
+logger = logging.getLogger(__name__)
+
+
+class TestPixelAttack(TestBase):
+    """
+    TODO: Write Comment
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        """
+        TODO: Write Comment
+        """
+        super().setUpClass()
+
+        cls.n_test = 2
+        cls.x_test_mnist = cls.x_test_mnist[0:cls.n_test]
+        cls.y_test_mnist = cls.y_test_mnist[0:cls.n_test]
+
+    def test_keras_mnist(self):
+        """
+        TODO: Write Comment
+        """
+        classifier = get_classifier_kr()
+        self._test_attack(
+            classifier,
+            self.x_test_mnist,
+            self.y_test_mnist,
+            False)
+
+    def test_tensorflow_mnist(self):
+        """
+        TODO: Write Comment
+        """
+        classifier, sess = get_classifier_tf()
+        self._test_attack(
+            classifier,
+            self.x_test_mnist,
+            self.y_test_mnist,
+            False)
+
+    def test_pytorch_mnist(self):
+        """
+        TODO: Write Comment
+        """
+        x_test = np.swapaxes(self.x_test_mnist, 1, 3).astype(np.float32)
+        classifier = get_classifier_pt()
+        self._test_attack(classifier, x_test, self.y_test_mnist, False)
+
+    def test_keras_mnist_targeted(self):
+        """
+        TODO: Write Comment
+        """
+        classifier = get_classifier_kr()
+        self._test_attack(
+            classifier,
+            self.x_test_mnist,
+            self.y_test_mnist,
+            True)
+
+    def test_tensorflow_mnist_targeted(self):
+        """
+        TODO: Write Comment
+        """
+        classifier, sess = get_classifier_tf()
+        self._test_attack(
+            classifier,
+            self.x_test_mnist,
+            self.y_test_mnist,
+            True)
+
+    def test_pytorch_mnist_targeted(self):
+        """
+        TODO: Write Comment
+        """
+        x_test = np.swapaxes(self.x_test_mnist, 1, 3).astype(np.float32)
+        classifier = get_classifier_pt()
+        self._test_attack(classifier, x_test, self.y_test_mnist, True)
+
+    def _test_attack(self, classifier, x_test, y_test, targeted):
+        """
+        TODO: Write Comment
+        """
+        x_test_original = x_test.copy()
+
+        if targeted:
+            # Generate random target classes
+            class_y_test = np.argmax(y_test, axis=1)
+            nb_classes = np.unique(class_y_test).shape[0]
+            targets = np.random.randint(nb_classes, size=self.n_test)
+            for i in range(self.n_test):
+                if class_y_test[i] == targets[i]:
+                    targets[i] -= 1
+        else:
+            targets = y_test
+
+        for es in [0, 1]:
+
+            df = PixelAttack(classifier, th=64, es=es, targeted=targeted)
+            x_test_adv = df.generate(x_test_original, targets)
+
+            self.assertFalse((x_test == x_test_adv).all())
+            self.assertFalse((0.0 == x_test_adv).all())
```

Comment:
Please add a test for the expected actual values of `x_test_adv`. See example: https://github.com/IBM/adversarial-robustness-toolbox/blob/e6d0b53a16bca37712a78ad916018c67a5951abe/tests/attacks/test_fast_gradient.py#L149

Reply:
Actually, as the prediction is calculated by the Evolutionary Optimisation which is stochastic in nature. The expected prediction will differ from machine to machine. Therefore, I did not add the expected predicted soft-labels (logits).

Reply:
Updated.

Reply:
Please note that the Pixel Attack is a black-box attack which uses Differential Evolution or Covariance Matrix Adaptation Evolution Strategy optimisations. Both of these optimisations have a bit of stochastics, which is why the perturbation can be different due to which the expected soft-label vector which is hard to predict.
In my experiments we have found there exists multiple solutions (perturbations) and we do not differentiate between them. However, if there was only a single solution possible, in that case, we would have a definite `y_test_pred_expected` value for that solution. 

In a gist, there exist multiple solutions and we don't differentiate between them so there is no expectation for the predicted y_test.

Reply:
Ok, I see. The reason why I'm asking is that the current tests only check if the attack runs without error and that the adversarial examples are different from the benign samples. Would there be a way to test more accurately if the attack works as expected? for example the number of pixels that get changed, the maximum difference between benign and adversarial examples, etc.?

Reply:
I see now what you expected. Please allow me to explain in detail about the examples and then about more accurate tests. 

In the case of the Pixel Attack, the number of pixels to be changed is defined by the variable `th` which is a required parameter for the attack other we find the minimum pixel by searching. Here, as we explicitly, define the `th` values; therefore, the attacks are coded such a way that search occurs in the given define space and not anywhere else. Again the difference between the benign samples and adversarial samples vary because each pixel can have maximum perturbation of 255 or minimum disruption of 0. For example, if I attack a sample with `th=10`. The image is perturbed such a way that only `10` pixels are modified but again getting the L<sub>1</sub> or  L<sub>2</sub> distance would be in a range and will vary for sample to sample. 

In the case of the Threshold Attack, all pixels can be changed to a defined upper limit `th` which is a required parameter for the attack other we find the minimum pixel by searching. Here, as we explicitly, define the `th` values; therefore, the attacks are coded such a way that search occurs in the given define space and not anywhere else. Again the difference between the benign samples and adversarial samples vary because maybe all pixels change or none changes. For example, if I attack a sample with `th=1`. The image is perturbed such a way that the pixels are modified in a way that all pixels have a maximum L<sub>1</sub> of `th` but again getting the L<sub>1</sub> or  L<sub>2</sub> distance for the entire image (adversarial sample) would be in a range and will vary for sample to sample. 

I don't think we need extensive tests to verify for the `th` values for the attacks because the evolutionary optimisation handles it and it defined explicitly in the function `get_bounds` for both the attacks. 

I would suggest adding the tests for the optimisation techniques if you require extensive tests.
Because if, the attack is not working as expected. It would be the optimisation which fails. However, I haven't encountered a test or a situation where optimisation didn't work as expected yet the attacks were working. I would run into errors if the optimisation didn't work.

Reply:
Ok. My recommendation would still be to add more assertions in the tests to make sure that the attack keeps working as intended in the future, but I don't want to insist on it.

Reply:
I see, I will think of some more assertions and add it in future. 
Currently, I can't think of any.